### PR TITLE
Attempt to fully explain all the things deducing this.

### PR DESCRIPTION
**Pull request by brevzin**

### CODE REVIEW

File: deducing-this.md [hunk 1]
```diff
@@ -203,80 +202,168 @@ class optional {
...
+    void foo(Self&& this self, int i);
+};
+
+X x;
+x.foo(4); // Self deduces as X&
+std::move(x).foo(2); // Self deduces as X
+```
 
-## What does `this` mean in the body of a member function?
+Since deduction rules do not change, and the explicit object parameter is just deduced from the object the function is called on, this has an interesting effect, which can best be illustrated by the following example:
```
Comment (TartanLlama) [inline deducing-this.md]: Remove "just" (pet peeve :D)

Comment (TartanLlama) [inline deducing-this.md]: I think it would be best to say "this has the interesting effect of possibly deducing derived types" rather than just "interesting effect". Otherwise, readers don't know what they're looking for in the example.

File: deducing-this.md [hunk 2]
```diff
@@ -191,9 +190,9 @@ template <typename T>
 class optional {
     // ...
     template <typename Self>
-    decltype(auto) value(Self&& this self) {
+    decltype(auto) value(Self&& this) {
         if (o.has_value()) {
-            return std::forward<Self>(self).m_value;
+            return std::forward<std::like_t<Self, optional>>(*self).m_value;
```
Comment (TartanLlama) [inline deducing-this.md]: std::like_t isn't introduced until later.

File: deducing-this.md [hunk 3]
```diff
@@ -203,80 +202,168 @@ class optional {
...
+ 
+In our view, options 2 and 3 are too likely to be lead to programmer errors, even by experts. It would be very surprising if `f1` didn't return `B::i`, or if `f1` and `f2` behaved differently. Option 4 is interesting, but would be more verbose without clear benefit. 
 
-## What about pass-by-value member functions?
+We favor option 1, since `f1`, `f2` and `f4` all end up having behavior that is reasonable given the other rules for name lookup we already have. `f2` and `f4` returning different things might appear strange at first glance, but `f2` references `this->i` and `f4` references `self.i` - programmers would simply have to be aware that these may be different. Where this option could go wrong is `f3`, which looks like it yield an lvalue or rvalue reference to `B::i` but is actually casting `*this` to a derived object. But since this is ill-formed, this is a *compile* error rather than a runtime bug. Hence, not only do we get typically expected behavior, but we're also protected from a potential source of error by the compiler.
```
Comment (TartanLlama) [inline deducing-this.md]: I'd remove "simply" here.

File: deducing-this.md [hunk 4]
```diff
@@ -636,14 +714,14 @@ template <typename T>
 class optional {
   // ...
   template <typename Self>
-  constexpr auto&& operator*(Self&& this self) {
-    return forward<Self>(self).m_value;
+  constexpr auto&& operator*(Self&& this) {
+    return forward<like_t<Self, optional>>(*this).m_value;
```
Comment (TartanLlama) [inline deducing-this.md]: Why do we need to use the `like_t` version here?

File: deducing-this.md (hunk 2 above)
Comment (atomgalaxy) [reply deducing-this.md]: `*self` doesn't make sense - `self` is not a pointer.

We need `std::forward<std::like_t<Self, optional>>(*this).m_value` because otherwise, we might not be accessing the correct subobject. However, in this case, the optional is inferrable, and so my darling forward_like is a much better alternative:
```
std::forward_like<Self>(m_value)
```
This just transfers the qualifiers from Self to whatever the type of value is, no need for anything else.

File: deducing-this.md (hunk 1 above)
Comment (atomgalaxy) [reply deducing-this.md]: agreed.

File: deducing-this.md (hunk 4 above)
Comment (atomgalaxy) [reply deducing-this.md]: If anyone inherits from optional you want to be sure you're accessing the correct member.

File: deducing-this.md (hunk 2 above)
Comment (brevzin) [reply deducing-this.md]: I added a section in 2. that introduces `forward_like` and `like` and links to @atomgalaxy 's implementation.

**Comment by brevzin** [comment]
So, good with these changes? Anything else big that needs to happen? All I'm doing this weekend is bingewatching the Olympics.

**Comment by TartanLlama** [comment]
I'm good with them.

**Comment by atomgalaxy** [comment]
Yeah, let's merge.

**Comment by atomgalaxy** [comment]
I'm gonna go through all the text again today, just to make sure I have, and I can't really devote any time to this tomorrow.

It's been a great pleasure cooperating. Who'll do the honors of sending the final revision to Hal Finkel on Monday?

File: deducing-this.md [hunk 5]
```diff
@@ -175,116 +175,213 @@ There are many, many cases in code-bases where we need two or four overloads of
...
+<table>
+<tr><th>C++17</th><th>With Explicit Object</th></tr>
+<tr><td>
+```
+struct X {
+    // implicit object has type X&
+    void foo();
 
-The entries of this table should be read as if they are inside a class `X`:
+    // implicit object has type X&
```
Comment (atomgalaxy) [inline deducing-this.md]: typo: X const&

File: deducing-this.md [hunk 6]
```diff
@@ -175,116 +175,213 @@ There are many, many cases in code-bases where we need two or four overloads of
...
-<tr><td>`void f(X const& this self)`           </td><td>`void f() const&`           </td><td>             </td></tr>
-<tr><td>`void f(X const&& this self)`          </td><td>`void f() const&&`          </td><td>             </td></tr>
-<tr><td>`void f(X volatile this self)`         </td><td>currently not available     </td><td>   [value]   </td></tr>
-<tr><td>`void f(X volatile& this self)`        </td><td>`void f() volatile&`        </td><td>             </td></tr>
-<tr><td>`void f(X volatile&& this self)`       </td><td>`void f() volatile&`        </td><td>             </td></tr>
-<tr><td>`void f(X const volatile this self)`   </td><td>currently not available     </td><td>   [value]   </td></tr>
-<tr><td>`void f(X const volatile& this self)`  </td><td>`void f() const volatile&`  </td><td>             </td></tr>
-<tr><td>`void f(X const volatile&& this self)` </td><td>`void f() const volatile&&` </td><td>             </td></tr>
+    // explicit object, unnamed, has type X
+    void ex_baz(X this);
```
Comment (atomgalaxy) [inline deducing-this.md]: I'd add "has type X, copy or move-constructed from original object"

File: deducing-this.md [hunk 7]
```diff
@@ -175,116 +175,213 @@ There are many, many cases in code-bases where we need two or four overloads of
...
+struct B {
+    int i = 0;
+    
+    template <typename Self>
+    auto&& get(Self&& this self) { return self.i; }
+};
 
-This means that rewriting the function signature in a different style should not change the ABI of your class, and you should also be able to implement a member function that is forward-declared with one syntax using the other.
+struct D : B {
+    double i = 3.14;
```
Comment (atomgalaxy) [inline deducing-this.md]: // shadows B::i

File: deducing-this.md [hunk 8]
```diff
@@ -175,116 +175,213 @@ There are many, many cases in code-bases where we need two or four overloads of
...
-## How do the explicit `this`-designated parameter and the current, trailing *cv-ref qualifiers* interact?
+Since deduction rules do not change, and the explicit object parameter is deduced from the object the function is called on, this has the interesting effect of possibly deducing derived types, which can best be illustrated by the following example:
 
-Other than the pass-by-value member functions, which currently do not have syntax to represent them, the explicit `this` signatures are aliases for those with trailing *cv-ref qualifiers*. They stand for the very same functions.
+```
+struct B {
+    int i = 0;
+    
+    template <typename Self>
+    auto&& get(Self&& this self) { return self.i; }
```
Comment (atomgalaxy) [inline deducing-this.md]: // note self.i, not simply i or this->i

File: deducing-this.md [hunk 9]
```diff
@@ -175,116 +175,213 @@ There are many, many cases in code-bases where we need two or four overloads of
...
+ 
+In our view, options 2 and 3 are too likely to be lead to programmer errors, even by experts. It would be very surprising if `f1` didn't return `B::i`, or if `f1` and `f2` behaved differently. Option 4 is interesting, but would be more verbose without clear benefit. 
 
-No change to current rules. Currently, one cannot have different cv-ref versions of either, so you cannot designate any parameter with `this`.
+We favor option 1, since `f1`, `f2` and `f4` all end up having behavior that is reasonable given the other rules for name lookup we already have. `f2` and `f4` returning different things might appear strange at first glance, but `f2` references `this->i` and `f4` references `self.i` - programmers would have to be aware that these may be different. Where this option could go wrong is `f3`, which looks like it yield an lvalue or rvalue reference to `B::i` but is actually casting `*this` to a derived object. But since this is ill-formed, this is a *compile* error rather than a runtime bug. Hence, not only do we get typically expected behavior, but we're also protected from a potential source of error by the compiler.
```
Comment (atomgalaxy) [inline deducing-this.md]: which looks like it yield an lvalue (-> yields)

Comment (atomgalaxy) [inline deducing-this.md]: I love this paragraph. So much win.

File: deducing-this.md [hunk 10]
```diff
@@ -295,17 +392,20 @@ class string_builder {
...
 
 // this is optimally efficient as far as allocations go
-std::string const x = (string_builder{{"asdf"}} * 5).bop().bop();
+std::string const x = (string_builder{"asdf"} * 5).bop().bop();
 ```
 
-Of course, implementing this example with templated `this` member functions would have been slightly more efficient due to also saving on move constructions, but the by-value `this` usage makes for simpler code.
+In this example, `x` would hold the value `"asdfasdfasdfasdfasdfbopbop"`.
+
+Of course, implementing this example with templated explicit object parameter member functions would have been slightly more efficient due to also saving on move constructions, but the by-value usage makes for simpler code.
```
Comment (atomgalaxy) [inline deducing-this.md:413]: The by-value builder method is also the only way you can pass a function pointer and still be able to use it by either move or copy constructing the parameter once you type-erase it - you can't do that with templates.

Can't really figure out how to word that properly but it's basically "if you want to type-erase a builder interface, the only way to do it is with pass-by-value methods, since templates are not virtualizable.

File: deducing-this.md [hunk 11]
```diff
@@ -370,24 +458,25 @@ If references and pointers do not have the same representation for member functi
...
-## Is `auto&& this self` allowed in member functions as well as lambdas?
+Virtual member function templates are not allowed in the language today, and this paper does not propose a change from the current behavior for member functions with explicit object parameters either.
 
-Yes. `auto&& param_name` has a well-defined meaning that is unified across the language. There is absolutely no reason to make it less so.
+## Can `static` member functions have a `this` parameter?
+
+No. Static member functions currently do not have an implicit `this` parameter, and therefore have no reason to have an explicit one.
 
-## Impact on the Standard
+## Interplays with capturing `[this]` and `[*this]`
```
Comment (atomgalaxy) [inline deducing-this.md]: add: in lambdas. I've read this far and got confused for 3 seconds, and I'm one of the authors.

File: deducing-this.md [hunk 12]
```diff
@@ -370,24 +458,25 @@ If references and pointers do not have the same representation for member functi
...
 
-TBD: A bunch of stuff in section 8.1.5 [expr.prim.lambda].
+`this` just designates the parameter that is bound to the reference to the function object. It does not, in any way, change the meaning of `this`.
```
Comment (atomgalaxy) [inline deducing-this.md]: *In the parameter list*, `this` just designates... ... change the meaning of `this` *in the body of a function*.

File: deducing-this.md [hunk 13]
```diff
@@ -370,24 +458,25 @@ If references and pointers do not have the same representation for member functi
...
 
-TBD: A bunch of stuff in that \this can appear as the first member function
-parameter.
+If other language features play with what `this` means, they are completely orthogonal and do not have interplays with this proposal. However, it should be obvious that develpers have a great potential for introducing hard-to-read code if they are at all changing the meaning of `this`, especially in conjunction with this proposal.
```
Comment (atomgalaxy) [inline deducing-this.md]: means *in the function body*

File: deducing-this.md [hunk 14]
```diff
@@ -441,7 +530,7 @@ This overload is callable for all `const`- and ref-qualified object parameters,
```
Comment (atomgalaxy) [inline deducing-this.md:569]: change:
the `volatile` versions are unlikely to be *valid*

File: deducing-this.md [hunk 15]
```diff
@@ -441,7 +530,7 @@ This overload is callable for all `const`- and ref-qualified object parameters,
 
 ## Alternative syntax
 
-Instead of qualifiying the parameter with `this`, to mark it as the parameter for deducing `this`, we could let the first parameter be named `this` instead:
+Instead of qualifying the parameter with `this`, to mark it as the parameter for deducing `this`, we could let the first parameter be *named* `this` instead:
```
Comment (atomgalaxy) [inline deducing-this.md:572]: However, that leaves us at a loss as to how one can address the derived object and the base object, as outlined in the previous section.

File: deducing-this.md [hunk 16]
```diff
@@ -747,6 +836,106 @@ Keep in mind that there are a few more functions in P0798 that have this lead to
...
+        [](auto const& this self, Node const& n) {
+            return std::visit(self, n.left) + std::visit(self, n.right);
+        }
+    ), tree);
+}
+```
+
+In this example, `self` doesn't deduce as that inner lambda, it deduces at the overload object, so this works.
+
+## CRTP, without the C, R, or even T
```
Comment (atomgalaxy) [inline deducing-this.md:930]: This made me emit a high-pitched noise that startled my wife. I demand forewarning for humor.

File: deducing-this.md (hunk 10 above)
Comment (brevzin) [reply deducing-this.md:413]: I'm not sure I understand this. Can you give me an example?

Comment (brevzin) [reply deducing-this.md:413]: Pushed corrections for all the comments but this one.

Comment (atomgalaxy) [reply deducing-this.md:413]: Lemme see:

```cpp
struct builder {
    std::string buffer;
    builder beep(builder this self) { return {buffer + "beep"}; }
    builder boop(builder this self) { return {buffer + "boop"}; }
};
using m_type = decltype(builder::beep);
static_assert(std::is_same(m_type, decltype(builder::boop)));

int boop_the_builder(m_type builder_xform) {
    // imagine this is shared_ptr<builder> and do_i_own is "ref-count is 1"
    auto&& [do_i_own, this_builder] = get_builder();
    if (do_i_own) {
        return builder_xform(std::move(this_builder));
    } else {
        return builder_xform(this_builder); // copy
    }
}

int main() {
    std::cout << boop_the_builder(builder::beep);
    std::cout << boop_the_builder(builder::boop);
}
```

Something like that.

 Basically the idea is that you don't know whether you'll own the object you'll feed the method that is *type-erased* - therefore, you got it, but it's most def not a template at this point - but you still only get to have one method, not a whole overload set. In those cases get-by-value is the only real solution, with move-into-parameter being the only proper way of handling it. Also I should probably have used std::invoke there but I really just don't have the energy.

**Comment by atomgalaxy** [comment]
Barry,

I left an explanation with an example for the bit that you asked about. I think this is all the energy I have left. Sorry - I've had a very, very demanding couple of weeks, with 2 weeks of 12-16h days rushing to get a demo of a product done, and then this week my boss/manager was on vacation so, being the lead developer, I had to pick up his workload of managing the rest of the people on our team who are in a different timezone. Suffice to say I'm dead.

So, basically, I want to say that I'm *really* grateful that you wrote all this stuff. It's unbelievably good - and as you saw, I read pretty thoroughly. You're my hero, man.

Thanks again for all your work. We all worked hard, but you're integrating all our work with completely new prose, which is another level.

I'm so tired I don't know how to end. So... Thanks. When I figure it out I'll post back :)

**Comment by atomgalaxy** [comment]
Yay!
